Performance investigation: Gunicorn worker and thread configurations (#74)
ONSdigital/eq-survey-runner-benchmark: doc/performance-investigations/0013-gunicorn-worker-thread-config/summary.md

This document highlights the results from running the application with different numbers of workers and thread permutations. Both workers and threads ranged from 1 to 10. 

## Benchmark profile

| Option                             | Value                        |
|------------------------------------|------------------------------|
| Requests file                      | census_household_gb_eng.json |
| Run time                           | 20m                          |
| User wait time minimum             | 1                            |
| User wait time maximum (exclusive) | 3                            |
| Clients                            | 225                          |
| Hatch rate                         | 225                          |
| Number of workers                  | 1 to 10                      |
| Number of threads                  | 1 to 10                      |
| Questionnaire Runner version       | v3.56.0                      |
| Questionnaire Runner min/min pods  | 3                            |

> | Max nodes per zone                 | 1                            |

Author: This is to highlight that 3 pods were between 3 nodes ensuring the kubesystem pods had to be one of the 3 nodes to ensure consistent between runs.

#### Workers: 6

![](images/performance_graph_workers6.png)

#### Workers: 7

![](images/performance_graph_workers7.png)

#### Workers: 8

![](images/performance_graph_workers8.png)

#### Workers: 9

![](images/performance_graph_workers9.png)

#### Workers: 10

![](images/performance_graph_workers10.png)

#### Comparison of the 99th percentile between worker and thread counts

![](images/performance_graph_workers_by_threads.png)

### Comparison of the 99th percentile between workers with a thread count of 2

The performance of the app does not seem to get noticeably better past 2 threads. The graph below highlights the performance of different workers under 2 threads over 3 runs.

![](images/performance_graph_workers_by_2_threads_99th_avg.png)

## Observations

- Using a thread count of 1 or worker count of 1 or 2 is not feasible.
- Our current configuration of 7 workers by 7 threads is not the most optimal.
- Under this test scenario, having more than 2 threads did not have any noticeable advantages in response times.
    - Along with this, having more than 4 workers does not seem to have any significant improvement in response times.

## Decision

> - Since different configuration has different pros and cons for different percentiles a few different worker/thread configurations will need further testing.
> - Run a small stress test to investigate the impact of thread counts when the app is under stress.
>     - Stress the app while using 7 workers with 7 threads vs 7 workers with 2 threads until the 99th percentile reaches 1 second. Following on from that, run a small stress test with reduced number of workers.

Reviewer: I don't think the first point is a decision, rather it leads into the decisions that we should run some stress tests:
```suggestion
- Since different configurations have different pros and cons for different percentiles, a few different worker/thread configurations should be tested further in small scale stress tests:
  - Compare a baseline of 7 workers with 7 threads vs 7 workers with 2 threads, until the 99th percentile reaches 1 second. 
  - Following on from that, run a test with a reduced number of workers.
```
